File: csharp/TheFactory.DatastoreTests/src/BatchTests.cs
using System;
using NUnit.Framework;
using TheFactory.Datastore;

namespace TheFactory.DatastoreTests {
    [TestFixture]
    public class BatchTests {
        [Test]
        public void TestBatchWriter() {
            Batch batch = new Batch();
            Assert.IsTrue(batch.IsEmpty());

            Slice[] pairs = new Slice[] {
                Utils.Slice("foo"), Utils.Slice("bar"),
                Utils.Slice("baz"), Utils.Slice("quux")
            };

            int i;
            for (i = 0; i < pairs.Length; i += 2) {
                batch.Put(pairs[i], pairs[i + 1]);
            }

            i = 0;
            foreach (IKeyValuePair kv in batch.Pairs()) {
                Assert.IsTrue(pairs[i].Equals(kv.Key));
                Assert.IsTrue(pairs[i + 1].Equals(kv.Value));
                Assert.IsFalse(kv.IsDeleted);
                i += 2;
            }
        }

        [Test]
        public void TestBatchWithDeletes() {
            Batch batch = new Batch();
            Assert.IsTrue(batch.IsEmpty());

            Slice[] pairs = new Slice[] {
                Utils.Slice("foo"), Utils.Slice("bar"),
                Utils.Slice("baz"), null,
                Utils.Slice("quux"), Utils.Slice("quuux")
            };

            int i;
            for (i = 0; i < pairs.Length; i += 2) {
                if (pairs[i + 1] == null) {
                    batch.Delete(pairs[i]);
                } else {
                    batch.Put(pairs[i], pairs[i + 1]);
                }
            }

            i = 0;
            foreach (IKeyValuePair kv in batch.Pairs()) {
                Assert.IsTrue(pairs[i].Equals(kv.Key));

                if (pairs[i + 1] == null) {
                    Assert.IsTrue(kv.IsDeleted);
                } else {
                    Assert.IsTrue(pairs[i + 1].Equals(kv.Value));
                    Assert.IsFalse(kv.IsDeleted);
                }
                i += 2;
            }
        }
    }
}

File: java/src/test/java/com/thefactory/datastore/BatchTests.java
package com.thefactory.datastore;

import junit.framework.TestCase;
import java.util.Iterator;

public class BatchTests extends TestCase {
        
    public void testBatchWriter() throws Exception {
        Batch batch = new Batch();
        assertTrue(batch.isEmpty());

        Slice[] pairs = new Slice[] {
            new Slice("foo".getBytes("UTF-8")), new Slice("bar".getBytes("UTF-8")),
            new Slice("baz".getBytes("UTF-8")), new Slice("quux".getBytes("UTF-8"))
        };

        int i;
        for (i = 0; i < pairs.length; i += 2) {
            batch.put(pairs[i], pairs[i + 1]);
        }

        Iterator<KV> kvs = batch.pairs();
        
        i = 0;
        while(kvs.hasNext()) {
          KV kv = kvs.next();
          assertEquals(kv.getKey(), pairs[i++]);
          assertEquals(kv.getValue(), pairs[i++]);
        }
        assertEquals(i, pairs.length);
    }

    public void testBatchWithDeletes() throws Exception {
        Batch batch = new Batch();
        assertTrue(batch.isEmpty());

        Slice[] pairs = new Slice[] {
            new Slice("foo".getBytes("UTF-8")), new Slice("bar".getBytes("UTF-8")),
            new Slice("baz".getBytes("UTF-8")), null,
            new Slice("quux".getBytes("UTF-8")), new Slice("quuux".getBytes("UTF-8"))
        };


        int i;
        for (i = 0; i < pairs.length; i += 2) {
            if (pairs[i + 1] == null) {
                batch.delete(pairs[i]);
            } else {
                batch.put(pairs[i], pairs[i + 1]);
            }
        }

        Iterator<KV> kvs = batch.pairs();

        i = 0;
        while(kvs.hasNext()) {
          KV kv = kvs.next();
          if(pairs[i + 1] == null) {
              assertEquals(kv.getKey(), pairs[i]);
              assertTrue(kv.isDeleted());
          } else {
              assertEquals(kv.getKey(), pairs[i]);
              assertEquals(kv.getValue(), pairs[i + 1]);
              assertFalse(kv.isDeleted());
          }
          i += 2;
        }
        assertEquals(i, pairs.length);
    }
}
